### PR TITLE
develop/check-ports-node-start

### PR DESCRIPTION
**Pull request by WinnieXU1**
1. check VNC port, ipmi_console_ssh_port, ipmi_console_port, bmc_connection_port and monitor port before node start.
2. add the tests

### CODE REVIEW

**Comment by fub2** [comment]
:+1: Comprehensive work!

File: test/functional/test_node_ports.py [hunk 1]
```diff
@@ -144,3 +145,303 @@ def test_start_node1_then_start_node2_no_conflict_ports(self):
...
+                'server': True,
+                'wait': False
+            }
+        }
+
+        try:
+            node2 = model.CNode(self.node_info_2)
+            node2.init()
+            node2.precheck()
+            node2.start()
```
Comment (turtle-fly) [inline test/functional/test_node_ports.py]: This exception should be raised in `precheck()`, so `start()` can be omitted here. 
Then I see your below tests to check there should be no active process.

Good job!

File: test/functional/test_node_ports.py [hunk 2]
```diff
@@ -144,3 +145,303 @@ def test_start_node1_then_start_node2_no_conflict_ports(self):
         assert "test1" in ipmi_result and "test2" in ipmi_result
         assert "test1-node" in qemu_result and "test2-node" in qemu_result
 
+    def test_start_node1_then_start_node2_bmc_connection_port(self):
```
Comment (turtle-fly) [inline test/functional/test_node_ports.py]: How about `test_start_nodes_with_conflict_bmc_connection_port()`, better then the foolish name I used in command line test.

So are other tests.

File: test/functional/test_node_ports.py [hunk 3]
```diff
@@ -144,3 +145,303 @@ def test_start_node1_then_start_node2_no_conflict_ports(self):
...
+                                   subprocess.PIPE, subprocess.PIPE)[1]
+        ipmi_result = run_command(PS_IPMI, True,
+                                  subprocess.PIPE, subprocess.PIPE)[1]
+        qemu_result = run_command(PS_QEMU, True,
+                                  subprocess.PIPE, subprocess.PIPE)[1]
+        "test2" not in socat_result
+        "test2" not in ipmi_result
+        "test2-node" not in qemu_result
+
+    def test_start_node1_then_start_node2_VNC_port(self):
```
Comment (turtle-fly) [inline test/functional/test_node_ports.py]: Considering test efficiency here, I have a better idea.

You can make all these tests into a class, in the class setup, you start `node-1`, and destroy it in class teardown. Every test try to start `node-2` using a conflict port. This method makes you start `node-1` only once and save quite a lot of time.

File: infrasim/model.py [hunk 4]
```diff
@@ -1225,6 +1226,32 @@ def precheck(self):
...
+                else:
+                    ssh_port = 9300
+        else:
+            ssh_port = 9300
+        if helper.check_if_port_in_use("0.0.0.0", ssh_port):
+            raise ArgsNotCorrect("ipmi console ssh port {} is already in use.".format(ssh_port))
+
+
+        # check if Monitor port is in use
+        if 'monitor' in self.__compute:
```
Comment (turtle-fly) [inline infrasim/model.py]: You can delegate `CMonitor` to do this precheck.

File: infrasim/model.py [hunk 5]
```diff
@@ -1225,6 +1226,32 @@ def precheck(self):
         if helper.check_if_port_in_use("0.0.0.0", self.__display + 5900):
             raise ArgsNotCorrect("VNC port {} is already in use.".format(self.__display + 5900))
 
+        # check if ipmi_console_ssh port is in use
+        infrasim_yml_path = os.path.join(self.get_workspace(), "etc", "infrasim.yml")
+        if os.path.exists(infrasim_yml_path):
+            with open(infrasim_yml_path, 'r') as fp:
+                conf = fp.read()
+                p_port = re.compile(r"^\s*ipmi_console_ssh:\s*(?P<port_ssh_for_client>\d+)",
```
Comment (turtle-fly) [inline infrasim/model.py]: Check `ipmi-console` port conflict in `console.py` or the package `ipmicons`.

**Comment by turtle-fly** [comment]
retest this please

**Comment by XiaowenJiang** [comment]
retest this please

**Comment by turtle-fly** [comment]
@WinnieXU1 sorry to be late, but infrasim start is much safer now!

**Pull request merged**